modules/gdscript/README.md

```diff
@@ -0,0 +1,111 @@
+
+The parser class also defines all the possible nodes of the AST, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class, or `GDScriptParser::FunctionNode` for a function. It further defines typechecking datastructures like `GDScriptParser::Datatype`.
+
+The parser was [intentionally designed](https://godotengine.org/article/gdscript-progress-report-writing-new-parser/#less-lookahead) with a look-ahead of a single token. This means that the parser only has access to the current token and the previous token (or, if you prefer, the current token and the next token). This parsing limitation ensures that GDScript will remain syntactically simple and accessible, and that the parsing process cannot become overly complex.
+
+
+### Analysis and typechecking (see [`GDScriptAnalyzer`](gdscript_analyzer.h))
+
+The analyzer takes in the AST of a program and verifies that "everything checks out". For example, when analyzing a method call with three parameters, it will check whether the function definition also contains three parameters. If the code is typed, it will check that the types are compatible.
+
+The analyzer is where a lot of the "magic" happens! It defines what constitutes proper code that can actually be compiled or run.
+
+This is also where dependencies are introduced. If a class `A` extends a class `B` or contains a member with type `B` from some other script file, then the analyzer will attempt to load that second script. If `B` contains references to `A`, then a _cyclic_ dependency is introduced.
+
+
+#### Cyclic dependencies and member resolution
+
+Cyclic dependencies from inheritance (`A extends B, B extends A`) are not supported in any programming language. Other cyclic dependencies are supported, such as `A extends B` and `B` contains, or preloads, members of type `A`. This is possible because of two features of the analyzer:
+
+1. Class interface resolution: to know what is available from one class, you don't need to analyze it in its entirety. You may merely analyze its "interface", meaning its name and its members (and their types!), as well as name and members of its base case, recursively. Details can be found in `GDScriptAnalyzer::resolve_class_interface()`.
+2. Out of order member resolution: the analyzer may not need an entire class interface to be resolved in order to figure out a specific type. For example, if class `A` contains code that references `B.is_alive`, then the analyzer doesn't need to immediately resolve `B`'s interface. It may simply check whether `is_alive` exists in `B`. This pattern happens often in the analyzer, and can be easily identified by the declaration of a temporary datatype with `GDScriptParser::DataType resolving_datatype;`, followed by `resolving_datatype.kind = GDScriptParser::DataType::RESOLVING;`. If the analyzer attempts to resolve a member that is already resolving, then a cyclic dependency problem has been found.
```

Comment:
Up for discussion, I can see both being accepted. The class is `DataType` so two words, but the local itself is `resolving_datatype`.
```suggestion
2. Out of order member resolution: the analyzer may not need an entire class interface to be resolved in order to figure out a specific type. For example, if class `A` contains code that references `B.is_alive`, then the analyzer doesn't need to immediately resolve `B`'s interface. It may simply check whether `is_alive` exists in `B`. This pattern happens often in the analyzer, and can be easily identified by the declaration of a temporary data type with `GDScriptParser::DataType resolving_datatype;`, followed by `resolving_datatype.kind = GDScriptParser::DataType::RESOLVING;`. If the analyzer attempts to resolve a member that is already resolving, then a cyclic dependency problem has been found.
```

Reply:
If it's OK, I think I prefer the way it is right now since like you mention it is `resolving_datatype` and not `resolving_data_type`.  This way we don't have to worry about writing data type vs datatype depending on context!